V2.0.9 restapi [FR]

alpes214 opened the pull request:
# Restapi endpoint
Restapi endpoint feature allows users to create new RESTapi endpoints and execute scripts on behalf of the proxysql. At the moment there are two examples of the scripts that are used for querying memory metrics and exporting users from MySQL database to ProxySQL.

## Installing dependencies
The python package `mysqlclient` must be installed in order `export_users` package to work. This can be accomplished by the `pip` command:
```
$ pip install mysqlclient
```
or, if proxysql is running under sudo:
```
$ sudo pip install mysqlclient
```

## Creating new endpoint
In order to create a new RESTapi endpoint, it must be defined in the `restapi_routes` table.
```
MySQL [(none)]> select * from restapi_routes;
Empty set (0.00 sec)

MySQL [(none)]> insert into restapi_routes (active, interval_ms, method, uri, script, comment) values (1,1000,'POST','export_users','./scripts/export_users.py','comm');
Query OK, 1 row affected (0.00 sec)

MySQL [(none)]> insert into restapi_routes (active, interval_ms, method, uri, script, comment) values (1,1000,'POST','metrics','./scripts/metrics.py','comm');
Query OK, 1 row affected (0.00 sec)

MySQL [(none)]> insert into restapi_routes (active, interval_ms, method, uri, script, comment) values (1,1000,'GET','metrics','./scripts/metrics.py','comm');
Query OK, 1 row affected (0.00 sec)

MySQL [(none)]> select * from restapi_routes\G
*************************** 1. row ***************************
         id: 1
     active: 1
interval_ms: 1000
     method: POST
        uri: export_users
     script: ./scripts/export_users.py
    comment: comm
*************************** 2. row ***************************
         id: 2
     active: 1
interval_ms: 1000
     method: POST
        uri: metrics
     script: ./scripts/metrics.py
    comment: comm
*************************** 3. row ***************************
         id: 3
     active: 1
interval_ms: 1000
     method: GET
        uri: metrics
     script: ./scripts/metrics.py
    comment: comm
3 rows in set (0.00 sec)

```

## Calling `metrics` example POST
Parameters to script are passed using json `-d '{"user":"root", "password":"a", "host":"127.0.0.1", "port":"6032”}'`
```
$ curl -X POST -d '{"user":"root", "password":"a", "host":"127.0.0.1", "port":"6032"}' http://127.0.0.1:6070/sync/metrics

{"params":{"user":"root", "password":"a", "host":"127.0.0.1", "port":"6032"}, "result":"Variable_Name\tVariable_Value\nSQLite3_memory_bytes\t3553504\njemalloc_resident\t21569536\njemalloc_active\t14565376\njemalloc_allocated\t10626296\njemalloc_mapped\t84004864\njemalloc_metadata\t5241896\njemalloc_retained\t51785728\nAuth_memory\t4042\nquery_digest_memory\t11832\nmysql_query_rules_memory\t1380\nmysql_firewall_users_table\t0\nmysql_firewall_users_config\t0\nmysql_firewall_rules_table\t0\nmysql_firewall_rules_config\t329\nstack_memory_mysql_threads\t33554432\nstack_memory_admin_threads\t16777216\nstack_memory_cluster_threads\t0\n"}
```

## Calling `metrics` example using method GET
Parameters to script are passed in the URL
```
val@s89830:~/workspace/val214_proxysql$ curl -X GET "http://127.0.0.1:6070/sync/metrics?user=root&password=a&host=127.0.0.1&port=6032"

{"params":{"host":"127.0.0.1","port":"6032","user":"root","password":"a"}, "result":"Variable_Name\tVariable_Value\nSQLite3_memory_bytes\t3157344\njemalloc_resident\t16977920\njemalloc_active\t12062720\njemalloc_allocated\t9356864\njemalloc_mapped\t67137536\njemalloc_metadata\t4779336\njemalloc_retained\t43487232\nAuth_memory\t2810\nquery_digest_memory\t0\nmysql_query_rules_memory\t1380\nmysql_firewall_users_table\t0\nmysql_firewall_users_config\t0\nmysql_firewall_rules_table\t0\nmysql_firewall_rules_config\t329\nstack_memory_mysql_threads\t33554432\nstack_memory_admin_threads\t16777216\nstack_memory_cluster_threads\t0\n"}
```
## Calling `export_user` example
The `export_users` script returns number of records processed in the json response.
```
$ curl -X POST -d '{"db":{"user":"root", "password":"a", "port":"3306", "host":"127.0.0.1"},"admin":{"user":"admin","password":"admin","port":"6032","host":"127.0.0.1"}}' http://127.0.0.1:6070/sync/export_users

{"params":{"db":{"user":"root", "password":"a", "port":"3306", "host":"127.0.0.1"},"admin":{"user":"admin","password":"admin","port":"6032","host":"127.0.0.1"}}, "result":{"num_records":"8"}}
```

Commit:
Implement call to external script and return stdout in response

Commit:
Implement restapi functionality
- restapi_routes table in admin module
- calling python script (json validation, error handling)
- scripts: metrics, export_users
- tests adjusted

yakirgb commented:
Great idea, thank U.

Commit:
Implementation of the GET and POST methods

Commit:
Use HTTP method for configuration

alikrubin commented:
Some minor issues:
Formatting error in the table:
1. 
```
Create Table: CREATE TABLE restapi_routes (
    id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1,
    interval_ms INTEGER CHECK (interval_ms>-100 AND interval_ms<=100000000) NOT NULL,
    method VARCHAR NOT NULL CHECK (UPPER(method) IN ('GET','POST')), uri VARCHAR NOT NULL, script VARCHAR NOT NULL, comment VARCHAR NO NULL DEFAULT '')
```
2. 
proxysql.cnf does not pickup the variable
```
mysql_variables=
{
    restapi_enabled="true"
}
```
```
MySQL [(none)]> select * from global_variables where variable_name like '%rest%';
+-----------------------+----------------+
| variable_name         | variable_value |
+-----------------------+----------------+
| admin-restapi_enabled | false          |
| admin-restapi_port    | 6070           |
+-----------------------+----------------+
2 rows in set (0.00 sec)
```
3. This section of config does not seem to work:
```
restapi =
(
	{
		id=1
		active=1
		interval_ms=1000
		uri="test"
		script="/script.py"
		comment="comment"
	}
)
```

alikrubin commented:
Another issue:
```
mysql> insert into restapi_routes (active, interval_ms, method, uri, script, comment) values (1,1000,'POST','export_users','./scripts/export_users.py','comm');
Query OK, 1 row affected (0.00 sec)

mysql> select * from restapi_routes;
+-----+--------+-------------+--------+--------------+---------------------------+---------+
| id  | active | interval_ms | method | uri          | script                    | comment |
+-----+--------+-------------+--------+--------------+---------------------------+---------+
| 101 | 1      | 1000        | POST   | export_users | ./scripts/export_users.py | comm    |
+-----+--------+-------------+--------+--------------+---------------------------+---------+
1 row in set (0.00 sec)

2020-01-11 23:08:21 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: table runtime_restapi_routes has 7 columns but 6 values were supplied --- INSERT INTO  runtime_restapi_routes  VALUES(1,1,1000,'POST','export_users','./scripts/export_users.py')
```

alikrubin commented:
Error
```
2020-01-11 23:14:59 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: table runtime_restapi_routes has 7 columns but 6 values were supplied --- INSERT INTO  runtime_restapi_routes  VALUES(1,1,1000,'POST','export_users','./scripts/export_users.py')
```

Commit:
Fix issue with method field

alpes214 commented:
@alikrubin thank you for your feedback.

> Some minor issues:
> Formatting error in the table:
> 1.
> 
> ```
> Create Table: CREATE TABLE restapi_routes (
>     id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
>     active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1,
>     interval_ms INTEGER CHECK (interval_ms>-100 AND interval_ms<=100000000) NOT NULL,
>     method VARCHAR NOT NULL CHECK (UPPER(method) IN ('GET','POST')), uri VARCHAR NOT NULL, script VARCHAR NOT NULL, comment VARCHAR NO NULL DEFAULT '')
> ```
> 
What do you mean with formatting error?

> 
> proxysql.cnf does not pickup the variable
> 
> ```
> mysql_variables=
> {
>     restapi_enabled="true"
> }
> ```
> 
> ```
> MySQL [(none)]> select * from global_variables where variable_name like '%rest%';
> +-----------------------+----------------+
> | variable_name         | variable_value |
> +-----------------------+----------------+
> | admin-restapi_enabled | false          |
> | admin-restapi_port    | 6070           |
> +-----------------------+----------------+
> 2 rows in set (0.00 sec)
> ```
> 
`admin-restapi_enabled` is an admin variable, but you are trying to configure it in the mysql_variables. I tried to configure
```
admin_variables=
{
    restapi_enabled=true
}
```
and it worked.

>     1. This section of config does not seem to work:
> 
> 
> ```
> restapi =
> (
> 	{
> 		id=1
> 		active=1
> 		interval_ms=1000
> 		uri="test"
> 		script="/script.py"
> 		comment="comment"
> 	}
> )
> ```

This is the section for automation tests. This is neither example nor correct configuration. It is used only for testing reading/writing configuration file.

alpes214 commented:
> Another issue:
> 
> ```
> mysql> insert into restapi_routes (active, interval_ms, method, uri, script, comment) values (1,1000,'POST','export_users','./scripts/export_users.py','comm');
> Query OK, 1 row affected (0.00 sec)
> 
> mysql> select * from restapi_routes;
> +-----+--------+-------------+--------+--------------+---------------------------+---------+
> | id  | active | interval_ms | method | uri          | script                    | comment |
> +-----+--------+-------------+--------+--------------+---------------------------+---------+
> | 101 | 1      | 1000        | POST   | export_users | ./scripts/export_users.py | comm    |
> +-----+--------+-------------+--------+--------------+---------------------------+---------+
> 1 row in set (0.00 sec)
> 
> 2020-01-11 23:08:21 sqlite3db.cpp:163:execute(): [ERROR] SQLITE error: table runtime_restapi_routes has 7 columns but 6 values were supplied --- INSERT INTO  runtime_restapi_routes  VALUES(1,1,1000,'POST','export_users','./scripts/export_users.py')
> ```

This is fixed. My bad.

I found one more problem which exists with `runtime_restapi_routes`. Currently user can use (call) RESTAPI service without loading configuration to runtime. This is not correct. I will fix it asap.

alikrubin commented:
@val214 thank you for the quick fix. I was talking about the different format (no end of lines after method field)  is due to the "space" is needed before and after "," in the table definition. Also there is a typo here: 
comment VARCHAR NO NULL DEFAULT '')
Probably needs to be 
comment VARCHAR NOT NULL DEFAULT '') 

Here is the diff (from the original version):
```
diff --git a/lib/ProxySQL_Admin.cpp b/lib/ProxySQL_Admin.cpp
index 7a677a3..b0bdb86 100644
--- a/lib/ProxySQL_Admin.cpp
+++ b/lib/ProxySQL_Admin.cpp
@@ -394,9 +394,9 @@ static int http_handler(void *cls, struct MHD_Connection *connection, const char
 
 #define ADMIN_SQLITE_TABLE_MYSQL_COLLATIONS "CREATE TABLE mysql_collations (Id INTEGER NOT NULL PRIMARY KEY , Collation VARCHAR NOT NULL , Charset VARCHAR NOT NULL , `Default` VARCHAR NOT NULL)"
 
-#define ADMIN_SQLITE_TABLE_RESTAPI_ROUTES "CREATE TABLE restapi_routes (id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT , active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1 , interval_ms INTEGER CHECK (interval_ms>-100 AND interval_ms<=100000000) NOT NULL , method VARCHAR NOT NULL CHECK (UPPER(method) IN ('GET','POST')), uri VARCHAR NOT NULL, script VARCHAR NOT NULL, comment VARCHAR NO NULL DEFAULT '')"
+#define ADMIN_SQLITE_TABLE_RESTAPI_ROUTES "CREATE TABLE restapi_routes (id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT , active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1 , interval_ms INTEGER CHECK (interval_ms>-100 AND interval_ms<=100000000) NOT NULL , method VARCHAR NOT NULL CHECK (UPPER(method) IN ('GET','POST')) , uri VARCHAR NOT NULL , script VARCHAR NOT NULL , comment VARCHAR NOT NULL DEFAULT '')"
 
-#define ADMIN_SQLITE_TABLE_RUNTIME_RESTAPI_ROUTES "CREATE TABLE runtime_restapi_routes (id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT , active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1 , interval_ms INTEGER CHECK (interval_ms>-100 AND interval_ms<=100000000) NOT NULL , method VARCHAR NOT NULL CHECK (UPPER(method) IN ('GET','POST')) , uri VARCHAR NOT NULL, script VARCHAR NOT NULL, comment VARCHAR NO NULL DEFAULT '')"
+#define ADMIN_SQLITE_TABLE_RUNTIME_RESTAPI_ROUTES "CREATE TABLE runtime_restapi_routes (id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT , active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1 , interval_ms INTEGER CHECK (interval_ms>-100 AND interval_ms<=100000000) NOT NULL , method VARCHAR NOT NULL CHECK (UPPER(method) IN ('GET','POST')) , uri VARCHAR NOT NULL , script VARCHAR NOT NULL , comment VARCHAR NOT NULL DEFAULT '')"
 
 #define ADMIN_SQLITE_TABLE_SCHEDULER "CREATE TABLE scheduler (id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL , active INT CHECK (active IN (0,1)) NOT NULL DEFAULT 1 , interval_ms INTEGER CHECK (interval_ms>=100 AND interval_ms<=100000000) NOT NULL , filename VARCHAR NOT NULL , arg1 VARCHAR , arg2 VARCHAR , arg3 VARCHAR , arg4 VARCHAR , arg5 VARCHAR , comment VARCHAR NOT NULL DEFAULT '')" 
```

alikrubin commented:
@val214 the new fix works just fine - thank you!
I would suggest also adding CORS headers and change content-type to json
cors header access-control-allow-origin 
Access-Control-Allow-Origin: *

Commit:
Fix problem with runtime table. Add CORS and content type headers

alpes214 commented:
@alikrubin changed formatting and added two headers. 

Also fixed problem with runtime table. Now restapi service will work if corresponding record exists in runtime_restapi_routes table.

alikrubin commented:
All is working fine now, thank you!
